nixos/make-options-doc: pretty-print literals

sandydoo opened the pull request:
###### Description of changes

Unlike the XML doc renderer, the AsciiDoc and CommonMark renderers don't pretty-print certain complex types, like literal expressions, DocBook literals, and derivations. These types are dumped into the documentation as JSON.

This commit parses and unwraps these types when loading the JSON-formatted NixOS options. The AsciiDoc and CommonMark renders have been combined into a single script to allow code reuse. The CommonMark renderer now also wraps the `type` attribute in a code block to match the formatting of the other attributes.

Here's an example from the [devenv](https://github.com/cachix/devenv) documentation, which happens to make use of the NixOS module documentation system:

```nix
{
  # ...
  package = mkOption {
    type = types.package;
    defaultText = literalExpression "pkgs.gradle.override { jdk = cfg.jdk.package; }";
    description = ''
      The gradle package to use.
      The gradle package by default inherits the JDK from `languages.java.jdk.package`.
    '';
  };
}
```

Rendering this to CommonMark, we see the following:

###### Before

    *_Default_*
    ```
    {"_type":"literalExpression","text":"pkgs.gradle.override { jdk = cfg.jdk.package; }"}
    ```

###### After

    *_Default_*
    ```
    "pkgs.gradle.override { jdk = cfg.jdk.package; }"
    ```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sandydoo committed:
nixos/make-options-doc: pretty-print literals

Unlike the XML doc renderer, the AsciiDoc and CommonMark renderers don't
pretty-print certain complex types, like literal expressions, DocBook
literals, and derivations. These types are dumped into the documentation
as JSON.

This commit parses and unwraps these types when loading the
JSON-formatted NixOS options. The AsciiDoc and CommonMark renders have
also been combined into a single script to allow code reuse.

domenkozar commented on nixos/lib/make-options-doc/generateDoc.py:
```diff
@@ -0,0 +1,106 @@
+        print(value['description'])
+        print()
+        if 'type' in value:
+            print('*_Type_*:')
```
Since we're changing this, could this match the output of the other outputs?

~~~
print('*__Type__*')
print ('```')
print(value['type'])
print ('```')
print ()

sandydoo committed:
nixos/make-options-doc: improve CommonMark formatting

Render the `type` attribute in a code block to match the rest of the
attributes.
